cpp/test/sg/fil_child_index_test.cu

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../../src/fil/internal.cuh"
+
+#include <test_utils.h>
+
+#include <cuml/fil/fil.h>
+#include <gtest/gtest.h>
+
+#include <cmath>
+#include <cstdio>
+#include <limits>
+#include <memory>
+#include <numeric>
+#include <ostream>
+#include <utility>
+
+namespace ML {
+
+using namespace fil;
+
+struct proto_inner_node {
+  bool def_left = false, is_categorical = false;
+  int fid      = 0;     // feature id
+  int set      = 0;     // which bit set represents the matching category list
+  float thresh = 0.0f;  // threshold
+  int left     = 1;     // left child idx
+  val_t split()
+  {
+    val_t split;
+    if (is_categorical)
+      split.idx = set;
+    else
+      split.f = thresh;
+    return split;
+  }
+  operator sparse_node16()
+  {
+    return sparse_node16({}, split(), fid, def_left, false, is_categorical, left);
+  }
+  operator sparse_node8()
+  {
+    return sparse_node8({}, split(), fid, def_left, false, is_categorical, left);
+  }
+  operator dense_node() { return dense_node({}, split(), fid, def_left, false, is_categorical); }
+};
+
+std::ostream& operator<<(std::ostream& os, const proto_inner_node& node)
+{
+  os << "def_left " << node.def_left << " is_categorical " << node.is_categorical << " fid "
+     << node.fid << " set " << node.set << " thresh " << node.thresh << " left " << node.left;
+  return os;
+}
+
+// proto inner node
+#define NODE(...)                                               \
+  []() {                                                        \
+    struct NonDefaultProtoInnerNode : public proto_inner_node { \
+      NonDefaultProtoInnerNode() { __VA_ARGS__; }               \
+    };                                                          \
+    return proto_inner_node(NonDefaultProtoInnerNode());        \
+  }()
+
+// proto category sets for one node
+struct ProtoCategorySets {
+  // each bit set for each feature id is in a separate vector
+  // read each uint8_t from right to left, and the vector(s) - from left to right
+  std::vector<std::vector<uint8_t>> bits;
```

Comment:
```suggestion
  std::vector<std::vector<uint8_t>> bits{};
```

Reply:
what changes when we use curly braces to invoke the default constructor?